### PR TITLE
V4.0.0 activation rules

### PR DESCRIPTION
**Pull request by kushti**

### CODE REVIEW

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: src/main/resources/mainnet.conf [hunk 1]
```diff
@@ -18,6 +18,15 @@ ergo {
     # Base16 representation of genesis state roothash
     genesisStateDigestHex = "a5df145d41ab15a01e0cd3ffbab046f0d029e5412293072ad0f5827428589b9302"
   }
+
+  voting {
+    # Activation height for mainnet protocol version 2 (client version 4.0.0 hard-fork)
+    version2ActivationHeight = 414720 // around Jan, 27th, 2021
+
+    # Difficulty for Autolykos version 2 activation (corresponding
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/resources/mainnet.conf:26]: fixed

File: src/main/scala/org/ergoplatform/settings/Parameters.scala [hunk 2]
```diff
@@ -127,6 +127,11 @@ class Parameters(val height: Height,
       table = table.updated(BlockVersion, table(BlockVersion) + 1)
       activatedUpdate = proposedUpdate
     }
+
+    if (height == votingSettings.version2ActivationHeight) {
+      require(table(BlockVersion) == 1, "Protocol version is not 1 on the hard-fork")
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/settings/Parameters.scala:135]: made a comment

**Review by aslesarenko — APPROVED** [review]
LGTM, but please address comments

File: src/main/scala/org/ergoplatform/mining/AutolykosSolution.scala [hunk 3]
```diff
@@ -75,3 +85,48 @@ object AutolykosSolutionSerializer extends ScorexSerializer[AutolykosSolution] {
...
+/**
+  * Binary serializer for Autolykos v2 solution, serializing and parsing "pk" and "nonce" values
+  */
+class AutolykosV2SolutionSerializer extends ScorexSerializer[AutolykosSolution] {
+
+  import AutolykosSolution.{wForV2, dForV2}
+
+  override def serialize(obj: AutolykosSolution, w: Writer): Unit = {
+    w.putBytes(groupElemToBytes(obj.pk))
+    w.putBytes(obj.n)
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/mining/AutolykosSolution.scala:97]: done

File: src/main/scala/org/ergoplatform/mining/AutolykosSolution.scala [hunk 4]
```diff
@@ -75,3 +85,48 @@ object AutolykosSolutionSerializer extends ScorexSerializer[AutolykosSolution] {
...
+
+}
+
+
+/**
+  * Serializing facade for both Autolykos v1 and v2 solutions
+  */
+object AutolykosSolutionSerializer {
+  val v1Serializer = new AutolykosV1SolutionSerializer
+  val v2Serializer = new AutolykosV2SolutionSerializer
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/mining/AutolykosSolution.scala:114]: done

File: src/main/scala/org/ergoplatform/mining/AutolykosSolution.scala [hunk 5]
```diff
@@ -75,3 +85,48 @@ object AutolykosSolutionSerializer extends ScorexSerializer[AutolykosSolution] {
...
+
+  private def serializer(blockVersion: Version): ScorexSerializer[AutolykosSolution] = {
+    if (blockVersion == 1) {
+      v1Serializer
+    } else {
+      v2Serializer
+    }
+  }
+
+  def serialize(h: Header, w: Writer): Unit = {
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/mining/AutolykosSolution.scala:124]: done

File: src/main/scala/org/ergoplatform/modifiers/history/PoPowAlgos.scala [hunk 6]
```diff
@@ -7,25 +7,29 @@ import scorex.util.{ModifierId, bytesToId, idToBytes}
...
 
 /**
   * A set of utilities for working with PoPoW security protocol.
   */
-object PoPowAlgos {
+class PoPowAlgos(powScheme: AutolykosPowScheme) {
 
   /**
-    * Computes interlinks vector for the next level after `prevHeader`.
+    * Computes max level (μ) of the given [[Header]], such that μ = log(T) − log(id(B))
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/history/PoPowAlgos.scala:21]: fixed

**Pull request merged**